### ISSUE TITLE
Updates not prevented when WooCommerce is outdated

### ISSUE THREAD

**Issue by adrianduffell** [Automattic/woocommerce-payments#1775]
In PR #1747 it was identified that bumping the `WC requires at least` value may have dangerous consequences for a site updating WC Pay but remaining on an old version of WooCommerce.

A potential solution was identified in https://github.com/Automattic/woocommerce-payments/pull/1747#issuecomment-836709557 

> The solution could be trying to hook into something like upgrader_pre_download or some other upgrader hook, fetch the updated plugin information, and stop the download/install before it happens. Interestingly, there aren't many resources on the internet that would describe how to solve such a scenario. If we were to work on this, the following test cases come to mind:

> auto update
> manual update
> manual bulk update with other plugins
> update on multisite

**Comment by marcinbot** [comment]
Another simple solution could be to just display a warning notice (non-dismissible maybe?) when an outdated dependency is detected, but not prevent the plugin from loading.

**Comment by adrianduffell** [comment]
> Another simple solution could be to just display a warning notice (non-dismissible maybe?) when an outdated dependency is detected, but not prevent the plugin from loading.

This sounds good. It would also be possible to add a note to the WooCommerce home screen and/or email via integrations with WC Admin.

So far I've hit a blocker stopping updates because the wordpress.org API doesn't return the `WC requires at least` value (see below). It could possibly be parsed out of the `readme.txt` description but that could be unreliable. I'm not sure if there is anywhere else the data could be fetched from?
___

https://api.wordpress.org/plugins/info/1.2/?action=plugin_information&request[slug]=woocommerce-payments

<img width="650" alt="Screen Shot 2021-05-13 at 11 31 25 am" src="https://user-images.githubusercontent.com/9312929/118073631-cb4f1000-b3de-11eb-9896-2b0fb5b81b74.png">

___

**Comment by adrianduffell** [comment]
Oops, I accidentally closed this 🤦 

>Another simple solution could be to just display a warning notice (non-dismissible maybe?) when an outdated dependency is detected, but not prevent the plugin from loading.

I agree I think with warning messages in place, we could remove the code which prevents the plugin loading when requirements are not met.

https://github.com/Automattic/woocommerce-payments/blob/670a492a071e96c865edb3f3b1e046cd75e3c807/includes/class-wc-payments.php#L137-L140

I think it's more preferable to let the plugin try to load so that the checkout has a chance to remain operational. This would then enable the `WC requires at least` to be set accurately (`check_plugin_dependencies` currently uses it).

**Comment by marcinbot** [comment]
I personally like the idea of loading the plugin even with the outdated dependencies. cc @jrodger for your input on this.

My only concern with this approach is if we were to start using some function in a critical filter that doesn't exist and then bring down the whole site on update. I think it was why the early return was put there. But we have already strayed a bit from doing critical actions only after that check, so I think we should be fine.

If we decide to go down this path, we could also try to break up the `check_plugin_dependencies` to clean up the code. The function is currently doing two things - return false if any dependency is outdated and render the notices if they are outdated and an optional flag is set. We could split it into functions like `is_woocommerce_updated`, `is_wc_admin_updated`, `is_wp_updated`, etc. (or similar names) and then use them to conditionally render the notices. We could even reuse the functions to add a box to the plugin page, similar to what Woo core is doing if it detects the plugin that hasn't been tested with it.

**Comment by jrodger** [comment]
> I personally like the idea of loading the plugin even with the outdated dependencies. cc @jrodger for your input on this.
>
>My only concern with this approach is if we were to start using some function in a critical filter that doesn't exist and then bring down the whole site on update. I think it was why the early return was put there. But we have already strayed a bit from doing critical actions only after that check, so I think we should be fine.

In principle I agree with loading the plugin with outdated dependencies, especially now that we're making the requirements stricter. Having a plugin completely turn off just because you dropped 3 versions behind on WooCommerce doesn't feel right.

I do wonder if we should have some sort of hard stop somewhere though? If developers can't rely on their being a certain version available it's going to be hard calling out to external dependencies. I think whatever we do we should create some sort of interface / wrapper with version checks for calling out to anything outside of WCPay (we're pretty good at doing this, but things like calling WC Admin Notes classes could be done better).

**Comment by marcinbot** [comment]
> I do wonder if we should have some sort of hard stop somewhere though?

I was thinking we could have such a mechanism as an extra measure. But by now I doubt anyone with an ancient version of WooCommerce will attempt to install WCPay after not having been able to do so for a year and seeing the minimum version header. So it's a nice-to-have, but probably not a blocker.

> I think whatever we do we should create some sort of interface / wrapper with version checks for calling out to anything outside of WCPay (we're pretty good at doing this, but things like calling WC Admin Notes classes could be done better).

I like this, especially that it would also potentially improve the code quality and make everything more unit-testable.

**Comment by adrianduffell** [comment]
> If developers can't rely on their being a certain version available it's going to be hard calling out to external dependencies.

The way I imagine it is if the installed WooCommerce is outside of L-2, there would be strong wording in a notice stating it needs to be updated immediately to prevent instability. So Ideally the site would only be in this state temporarily, and developers should be able to rely on L-2 as a normal baseline.

There is still a concern though that an outdated WooCommerce could result in a fatal error/white screen that prevents the warning notice from displaying. From that perspective, a hard stop + warning notice might be preferable. 

It can also be part of GlobalStep’s testing procedures to test the warning notice when an old WooCommerce is installed -- this is done for WC Admin presently.

>  think whatever we do we should create some sort of interface / wrapper with version checks for calling out to anything outside of WCPay (we're pretty good at doing this, but things like calling WC Admin Notes classes could be done better).

💯

**Comment by adrianduffell** [comment]
> If developers can't rely on their being a certain version available it's going to be hard calling out to external dependencies.

I think I misunderstood this in my message above. Yes, I agree - for example there might be a call to a Core class method needing to be made, but it's only available in L-2. So calls out to external dependencies need to be made defensively if there is no hard stop in place. I think your idea to add a safe interface / wrapper around the dependencies would be a great option.

**Comment by marcinbot** [comment]
With manual testing (or we could add an automated test for that even) we could just not overthink this and allow the plugin to always update, load, and just display notices on outdated dependencies.

The manual test step could be:
"On a site with WP 5.4 and WC 4.0 make sure that a notice about the outdated dependencies is shown, and that it's possible to deactivate the WCPay plugin"

I'm using the oldest dependencies we specify right now, and just want to make sure that the site doesn't get completely broken and can be fixed by deactivating the plugin. Everything else (potentially broken checkout or other pages) can be fixed by updating the dependencies.

**Comment by RadoslavGeorgiev** [comment]
Let's check if https://github.com/Automattic/woocommerce-payments/pull/3010 is a sufficient solution for this issue.

**Comment by vbelolapotkov** [comment]
@jrodger I see this issues is assigned to you and it has at least 4 merged PRs linked to it. Is there any other work expected?

**Comment by jrodger** [comment]
The approach we settled on in #3010 resolves this issue in the best way we have available to us at the moment. I'll document the current behaviour in our documentation in this repository so that we can refer to it in the future.

I'll keep this issue open for now and close it once I have that documentation PR merged.

**Comment by htdat** [comment]
@jrodger - I can continue this PR and document the current behavior. But can you let me know which documents did you want to update?

**Comment by jrodger** [comment]
Thanks @htdat! I was going to add a "page" (markdown file) detailing our version support policy here: https://github.com/Automattic/woocommerce-payments/tree/develop/docs. Perhaps linking to it from the main readme?

We've announced the L-2 policy on the WooCommerce blog and it's also communicated through the `readme.txt` attributes, so the document here would be more focused on WCPay developers (i.e. to help understand what features they can use / need to support and for how long).
